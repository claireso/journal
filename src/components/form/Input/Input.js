import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import { Group } from '../Group'
import Label from '../Label'

import * as S from './Input.styles'

const Input = React.forwardRef(({ onChange, name, ...props }, ref) => {
  const handleChange = useCallback(
    (event) => {
      onChange(name, event.target.value)
    },
    [onChange, name]
  )

  return (
    <Group>
      <Label htmlFor={name}>{props.label}</Label>
      <S.StyledInput
        ref={ref}
        id={name}
        data-testid={props.testId}
        type={props.type || 'text'}
        name={name}
        defaultValue={props.value}
        required={props.required}
        onChange={onChange && handleChange}
        autoFocus={props.autoFocus}
      />
    </Group>
  )
})

Input.displayName = 'Input'

Input.propTypes = {
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  type: PropTypes.string,
  required: PropTypes.bool,
  onChange: PropTypes.func,
  autoFocus: PropTypes.bool,
  testId: PropTypes.string
}

Input.defaultProps = {
  onChange: () => {}
}

export default Input
