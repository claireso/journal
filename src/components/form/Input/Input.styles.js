import styled from 'styled-components'

export const StyledInput = styled.input`
  appearance: none;
  background: var(--form-input-bg);
  border: 1px solid var(--border-color);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  display: block;
  font-family: var(--font-family);
  outline: none;
  padding: 0.8rem 1.2rem;
  width: 100%;
  font-size: var(--font-size-normal);
  line-height: 1.71;
  transition: border-color 250ms ease-out;
  &:focus {
    border-color: var(--border-color-focus);
  }
`
