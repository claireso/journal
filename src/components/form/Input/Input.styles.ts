import { styled } from '@theme'

export const StyledInput = styled('input', {
  appearance: 'none',
  background: '$white',
  border: '1px solid $gray200',
  borderRadius: '$1',
  boxShadow: '$2',
  display: 'block',
  fontFamily: '$sansSerif',
  outline: 'none',
  p: '$2 $3',
  width: '100%',
  fontSize: '$4',
  lineHeight: 1.71,
  transition: '$borderColor',
  '&:focus': {
    borderColor: '$gray300'
  }
})
