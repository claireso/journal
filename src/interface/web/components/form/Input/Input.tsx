'use client'
import React, { useCallback } from 'react'

import Group from '../Group'
import Label from '../Label'

import * as cls from './styles.css'

interface InputProps extends Omit<React.ComponentProps<'input'>, 'onChange'> {
  name: string
  label: string
  onChange?: (name: string, value: string) => void
}

const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const { onChange, name, label, value = '', type = 'text', required = false, autoFocus = false } = props

  const handleChange = useCallback(
    (event: React.FormEvent<HTMLInputElement>) => {
      onChange && onChange(name, event.currentTarget.value)
    },
    [onChange, name]
  )

  return (
    <Group>
      <Label htmlFor={name}>{label}</Label>
      <input
        ref={ref}
        className={cls.input}
        id={name}
        type={type}
        name={name}
        defaultValue={value}
        required={required}
        onChange={handleChange}
        autoFocus={autoFocus}
      />
    </Group>
  )
})

Input.displayName = 'Input'

export default Input
