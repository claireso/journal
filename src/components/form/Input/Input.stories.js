import React from 'react'
import { action } from '@storybook/addon-actions'

import Input from './index'

export default {
  title: 'Form/Input'
}

export const Basic = () => <Input label="Label" name="title" onChange={action('on change')} />
