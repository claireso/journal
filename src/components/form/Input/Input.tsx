import React, { useCallback } from 'react'

import { Group } from '../Group'
import Label from '../Label'

import * as S from './Input.styles'

interface InputProps {
  name: string
  label: string
  testId?: string
  type?: string
  value?: string
  required?: boolean
  autoFocus?: boolean
  onChange?: (name: string, value: string) => void
}

const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const { onChange, name, label, value = '', type = 'text', required = false, autoFocus = false, testId } = props

  const handleChange = useCallback(
    (event) => {
      onChange && onChange(name, event.target.value)
    },
    [onChange, name]
  )

  return (
    <Group>
      <Label htmlFor={name}>{label}</Label>
      <S.StyledInput
        ref={ref}
        id={name}
        data-testid={testId}
        type={type}
        name={name}
        defaultValue={value}
        required={required}
        onChange={handleChange}
        autoFocus={autoFocus}
      />
    </Group>
  )
})

Input.displayName = 'Input'

export default Input
