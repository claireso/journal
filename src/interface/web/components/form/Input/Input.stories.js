import React from 'react'
import { action } from '@storybook/addon-actions'

import Input from './index'

const params = {
  title: 'Form/Input'
}

export default params

export const Basic = () => <Input label="Label" name="title" onChange={action('on change')} />
