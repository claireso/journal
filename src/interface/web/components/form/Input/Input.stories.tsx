import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Input from './Input'

const meta: Meta<typeof Input> = {
  title: 'Form/Input',
  component: Input
}

export default meta
type Story = StoryObj<typeof Input>

export const Basic: Story = {
  args: {
    label: 'Label',
    name: 'title',
    onChange: action('on change')
  }
}
