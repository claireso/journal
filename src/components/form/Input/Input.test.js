import { render, fireEvent } from '@testing-library/react'

import Input from './index'

describe('<Input />', () => {
  const props = {
    label: 'My input',
    name: 'myinput',
    value: 'my custom value'
  }

  test('should render input', () => {
    const { container } = render(<Input {...props} />)

    expect(container).toMatchSnapshot()
  })

  test('should call onChange property', () => {
    const spyOnChange = jest.fn()

    const { getByLabelText } = render(<Input {...props} onChange={spyOnChange} />)

    const input = getByLabelText('My input')

    fireEvent.change(input, { target: { value: 'My new value' } })

    expect(spyOnChange).toHaveBeenCalledWith('myinput', 'My new value')
  })
})
