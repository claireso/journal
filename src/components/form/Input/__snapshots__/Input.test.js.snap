// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render input 1`] = `
.c0 {
  margin: 0 0 var(--gutter);
}

.c1 {
  display: block;
  font-size: 1.6rem;
  line-height: 1.5;
  margin: 0 0 1.2rem;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: var(--form-input-bg);
  border: 1px solid var(--border-color);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  display: block;
  font-family: var(--font-family);
  outline: none;
  padding: 0.8rem 1.2rem;
  width: 100%;
  font-size: var(--font-size-normal);
  line-height: 1.71;
  -webkit-transition: border-color 250ms ease-out;
  transition: border-color 250ms ease-out;
}

.c2:focus {
  border-color: var(--border-color-focus);
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      for="myinput"
    >
      My input
    </label>
    <input
      class="c2"
      id="myinput"
      name="myinput"
      type="text"
      value="my custom value"
    />
  </div>
</div>
`;
